File: dynamic_programming/03_pascal_triangle/pascal_triangle.py
from typing import List
from copy import copy


class Solution:
    def generate(self, num_rows: int) -> List[List[int]]:
        if not self._check_constraints(num_rows):
            raise ValueError('Constraints violated')
        ans: List[List[int]] = []
        dp = [1]
        ans.append(copy(dp))
        if num_rows > 1:
            dp.append(1)
            ans.append(copy(dp))
            for j in range(1, num_rows - 1):
                for i in range(j, 0, -1):
                    dp[i] += dp[i - 1]
                dp.append(1)
                ans.append(copy(dp))

        return ans

    def _check_constraints(self, n: int) -> bool:
        if n < 1 or n > 30:
            return False
        return True

File: dynamic_programming/03_pascal_triangle/pascal_triangle.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <cstdint>
#include <vector>

using namespace std;


class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        if (!checkConstraints(numRows)) {
            throw std::runtime_error("Constraints violated");
        }
        vector<vector<int>> ans;
        vector<int> dp(1, 1);
        ans.push_back(dp);
        if (numRows > 1) {
            dp.push_back(1);
            ans.push_back(dp);
            for (int j = 1; j < numRows - 1; j++) {
                for (int i = j; i > 0; i--) {
                    dp[i] += dp[i - 1];
                }
                dp.push_back(1);
                ans.push_back(dp);
            }
        }
        return ans;
    }

private:
    bool checkConstraints(int n) {
        if (n < 1 || n > 30) return false;
        return true;
    }
};
